compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -343,7 +343,12 @@ object TypeOps:
 
     tp match {
       case tp: OrType =>
-        approximateOr(tp.tp1, tp.tp2)
+        (tp.tp1.dealias, tp.tp2.dealias) match
+          case (tp1 @ AppliedType(tycon1, args1), tp2 @ AppliedType(tycon2, args2))
```

Comment:
While this addresses the most common issue, I don't think that's enough to handle things like `Tag[Long] | (Tag[Int] & SomeTrait)` or `type Tag2[X] <: Tag[X]; Tag[Long] | Tag2[Int]`. But to properly deal with all possible cases it seems that we'd need a generalization of `baseClasses` that also returns base abstract types.

Reply:
Since what we do here is an approximation anyway, I'd leave that under possible future improvement.

Reply:
Sounds good to me.